### ISSUE TITLE
Hot reload doesn't work in 4.x with ModuleFederationPlugin

### ISSUE THREAD

**Issue by alesculek** [webpack/webpack-dev-server#4378]
### Bug report

Webpack dev server doesn't perform the hot reload properly in conjunction with `ModuleFederationPlugin` in v4.x but works in v3.x

### Actual Behavior

When I set up a simple project using those dependencies:

```
{
    "name": "test",
    "version": "1.0.0",
    "scripts": {
        "start": "webpack serve --config=config/webpack.dev.js",
        "build": "webpack --config=config/webpack.prod.js"
    },
    "dependencies": {
        "react": "^17.0.1",
        "react-dom": "^17.0.1",
        "react-router-dom": "^5.2.0"
    },
    "devDependencies": {
        "@babel/core": "^7.17.9",
        "@babel/plugin-transform-runtime": "^7.17.0",
        "@babel/preset-env": "^7.16.11",
        "@babel/preset-react": "^7.16.7",
        "babel-loader": "^8.2.4",
        "clean-webpack-plugin": "^4.0.0",
        "html-webpack-plugin": "^5.5.0",
        "webpack": "^5.71.0",
        "webpack-cli": "^4.9.2",
        "webpack-dev-server": "^4.8.1",
        "webpack-merge": "^5.8.0"
    }
}
```

I manage to startup my application, I can't see any errors but when I update the source code, the update is not propagated into the browser and the change is visible ONLY after the full reload.

When I downgrade back to 3.x everything works smoothly.

I used this repository as reference: https://github.com/richardtbell/microfrontend-hello-world

Since this repository was quite older, I wanted to update to the latest version of all packages, the only issue I had was with the webpack-dev-server as I mentioned above.

### Expected Behavior

A SPA is HOT reloaded as expected

### How Do We Reproduce?

1. Clone https://github.com/richardtbell/microfrontend-hello-world
2. run: npm run start, update App.js, see the change in the browser is being propagated
3. Update the dependencies with the snippet above.
4. Do the same
5. The change won't be visible, only after reload
6. Try to downgrade the dev server to 3.x
7. Everything will work
8. OR: keep 4.x and remove the ModuleFederationPlugin from the webpack config and the hot reload will work again

### Please paste the results of `npx webpack-cli info` here, and mention other relevant information

```
❯ npx webpack-cli info

  System:
    OS: macOS 12.3
    CPU: (8) x64 Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
    Memory: 29.45 MB / 16.00 GB
  Binaries:
    Node: 14.17.2 - ~/.nvm/versions/node/v14.17.2/bin/node
    Yarn: 1.22.15 - ~/.yarn/bin/yarn
    npm: 8.6.0 - ~/.nvm/versions/node/v14.17.2/bin/npm
  Browsers:
    Brave Browser: 99.1.36.119
    Chrome: 100.0.4896.75
    Safari: 15.4
  Packages:
    babel-loader: ^8.2.4 => 8.2.4
    clean-webpack-plugin: ^4.0.0 => 4.0.0
    html-webpack-plugin: ^5.5.0 => 5.5.0
    webpack: ^5.71.0 => 5.72.0
    webpack-cli: ^4.9.2 => 4.9.2
    webpack-dev-server: ^4.8.1 => 4.8.1
    webpack-merge: ^5.8.0 => 5.8.0
```

**Comment by alesculek** [comment]
Update:

I'm also getting this error message:

```
Uncaught TypeError: Cannot set properties of undefined (setting './src/App.js')
    at self.webpackHotUpdatetest (remoteEntry.js:1155:39)
    at src_bootstrap_js.1a6c7b3ebe42c99bbf18.hot-update.js:10:29
```

This happens only if the App react component has other dependencies (in this case the history), if it does not, there is NO message in the logs.

If I open this in chrome, I can see those lines there:

```
**/ 		self["webpackHotUpdatetest"] = (chunkId, moreModules, runtime) => {
/******/ 			for(var moduleId in moreModules) {
/******/ 				if(__webpack_require__.o(moreModules, moduleId)) {
/******/ 					currentUpdate[moduleId] = moreModules[moduleId];
/******/ 					if(currentUpdatedModulesList) currentUpdatedModulesList.push(moduleId);
/******/ 				}
/******/ 			}
/******/ 			if(runtime) currentUpdateRuntime.push(runtime);
/******/ 			if(waitingUpdateResolves[chunkId]) {
/******/ 				waitingUpdateResolves[chunkId]();
/******/ 				waitingUpdateResolves[chunkId] = undefined;
/******/ 			}
/******/ 		};
```

where to problematic one seems to be:

```
currentUpdate[moduleId] = moreModules[moduleId];
```

**Comment by LadaBr** [comment]
I also experience this issue. It happens only when MFE exposes some script. Temporary solution is setting FAST_REFRESH=false in .env file. But page will be reloaded after each change.

**Comment by felixmokross** [comment]
We are experiencing the same issue.

Falling back to FAST_REFRESH=false, however is also not working for us. If I add the ModuleFederationPlugin, the page no longer reloads upon changes (although it shows the `Compiling...` message in the console).

But we anyway need hot module replacement as we have complex in-memory state which would get lost upon page reload.

**Comment by alexander-akait** [comment]
Anyone tried to debug this and find the problem?

**Comment by felixmokross** [comment]
I digged a bit deeper now. It's still not clear to me where the problem is exactly. I see that messages are exchanged via the websocket and I see that the modules still are loaded, even with the ModuleFederationPlugin enabled. When debugging, it seems to me that the modules are replaced (hard to judge for me though as I'm not familiar with the webpack dev server code base). When I check the module source in the browser, I also see the updated code.

I think the issue might be rather that React somehow doesn't manage to re-render the respective components. So the problem would rather be in the [React Refresh plugin](https://github.com/pmmmwh/react-refresh-webpack-plugin), which is used by CRA 5. But not yet sure about it. Will have another look tomorrow.

**Comment by felixmokross** [comment]
For us, it turned out to be this issue in the React Refresh webpack plugin, which comes with CRA: https://github.com/pmmmwh/react-refresh-webpack-plugin/issues/394 I don't see this plugin in the repro steps of the op, so this might be a separate issue.

Workaround for us is to opt-into Module Federation when starting the dev server (via a command line arg), so that most developers are not impacted by this when developing the standalone application. When consuming a component via Module Federation, Fast Refresh doesn't work anyway, so we are not losing anything.

**Comment by alexander-akait** [comment]
Try to enable https://webpack.js.org/configuration/optimization/#optimizationruntimechunk

**Comment by tony-show** [comment]
> Try to enable https://webpack.js.org/configuration/optimization/#optimizationruntimechunk

Thanks, but it doesn't work right.

This only works in microfronts where there are no remotes. But where remotes are specified, the page reloads infinitely. (video demo):
https://disk.yandex.ru/i/-E5A1LpdhNIn7w

**Comment by alexander-akait** [comment]
You have a lot of hot entrypoint on an one page, sounds like something wrong with your setup

**Comment by tony-show** [comment]
> You have a lot of hot entrypoint on an one page, sounds like something wrong with your setup

host - page reloads infinitely
**host: webpack.config.js**
`import path from 'path'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import webpack from 'webpack'
const ModuleFederationPlugin = webpack.container.ModuleFederationPlugin

export default {
  mode: 'development',
  entry: {
    main: path.resolve(__dirname, 'src', 'index.ts')
  },
  output: {
    filename: '[name].js',
    publicPath: "auto",
  },
  devServer: {
    port: 4000,
    open: true,
    historyApiFallback: true,
    hot: true,
  },
  optimization: {
    runtimeChunk: {
      name: 'runtime',
    },
  },
  module: {
    rules: [
      {
        test: /\.j|tsx?$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: [
              ["@babel/preset-typescript", {
                isTSX: true,
                allExtensions: true
              }],
            ],
          }
        }
      }
    ]
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.js', '.json'],
  },
  devtool: 'source-map',
  plugins: [
    new ModuleFederationPlugin({
      name: 'host',
      filename: "remoteEntry.js",
      remotes: {
        host: 'host@http://localhost:4000/remoteEntry.js',
        app1: 'app1@http://localhost:4001/remoteEntry.js',
        ui_kit: "ui_kit@http://localhost:4002/remoteEntry.js",
      },
      exposes: {
        './posts': './src/services/posts.ts',
        "./store": "./src/store/store.tsx",
      },
      shared: {
        'react-redux': {
          singleton: true,
        },
      }
    }),
    new HtmlWebpackPlugin({
      template: './public/index.html',
    }),
  ]
};
`

app1 - page reloads infinitely
**app1: webpack.config.js**
`import path from 'path'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import webpack from 'webpack'
const ModuleFederationPlugin = webpack.container.ModuleFederationPlugin

export default {
  mode: 'development',
  entry: {
    main: path.resolve(__dirname, 'src', 'index.ts')
  },
  output: {
    filename: '[name].js',
    publicPath: "auto",
  },
  devServer: {
    port: 4001,
    open: true,
    historyApiFallback: true,
    hot: true,
  },
  optimization: {
    runtimeChunk: {
      name: 'runtime',
    },
  },
  devtool: 'source-map',
  module: {
    rules: [
      {
        test: /\.j|tsx?$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: [
              ["@babel/preset-typescript", {
                isTSX: true,
                allExtensions: true
              }],
            ],
          }
        }
      }
    ]
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.js'],
  },
  plugins: [
    new ModuleFederationPlugin({
      name: 'app1',
      filename: "remoteEntry.js",
      remotes: {
        host: "host@http://localhost:4000/remoteEntry.js",
        ui_kit: "ui_kit@http://localhost:4002/remoteEntry.js",
      },
      exposes: {
        "./RemoteComponent": "./src/RemoteComponent.tsx",
      },
      shared: {
        'react-redux': {
          singleton: true,
        },
      }
    }),
    new HtmlWebpackPlugin({
      template: './public/index.html',
    }),
  ]
};
`

ui_kit - page works correctly with a hot reload
**ui_kit: webpack.config.js**
`import path from 'path'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import webpack from 'webpack'
const ModuleFederationPlugin = webpack.container.ModuleFederationPlugin

export default {
  mode: 'development',
  entry: {
    main: path.resolve(__dirname, 'src', 'index.ts')
  },
  output: {
    filename: '[name].js',
    publicPath: "auto",
  },
  devServer: {
    port: 4002,
    open: true,
    historyApiFallback: true,
    hot: true,
  },
  optimization: {
    runtimeChunk: {
      name: 'runtime',
    },
  },
  module: {
    rules: [
      {
        test: /\.j|tsx?$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: {
            presets: [
              ["@babel/preset-typescript", {
                isTSX: true,
                allExtensions: true
              }],
            ],
          }
        }
      }
    ]
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.js'],
  },
  plugins: [
    new ModuleFederationPlugin({
      name: 'ui_kit',
      filename: "remoteEntry.js",
      exposes: {
        "./components": "./src/components/index.ts",
      },
    }),
    new HtmlWebpackPlugin({
      template: './public/index.html',
    }),
  ]
};
`

**Comment by alexander-akait** [comment]
@tony-show Can you create a small reproducible example with your configuration and code (no nee to provide all code, only piece for an example), this will give me a better look

**Comment by tony-show** [comment]
> You have a lot of hot entrypoint on an one page, sounds like something wrong with your setup

https://github.com/tony-show/module-federation-example

**Comment by tony-show** [comment]
> @tony-show Can you create a small reproducible example with your configuration and code (no need to provide all code, only piece for an example), this will give me a better look

https://github.com/tony-show/module-federation-example

**Comment by alexander-akait** [comment]
@tony-show How you run it? You **can't** use `name: ""` for your packages, it should be always an unique name, shorty - it seems webpack is thinking your build is one (becaise chunk loading method is not unique), alternatively you can set `output.uniqueName`(https://webpack.js.org/configuration/output/#outputuniquename) for each compiler

**Comment by tony-show** [comment]
> @tony-show How you run it? You **can't** use `name: ""` for your packages, it should be always an unique name, shorty - it seems webpack is thinking your build is one (becaise chunk loading method is not unique), alternatively you can set `output.uniqueName`(https://webpack.js.org/configuration/output/#outputuniquename) for each compiler

After adding the uniqueName property with a unique name for each microfrontend application, those with the remotes setting in Module Federation now have a remoteEntry.js load error, although the file is available. Well and the page displays a white screen.

![image](https://user-images.githubusercontent.com/15953087/236762649-6d65d31d-d9e3-49af-b2c6-8c4628c64078.png)

![image](https://user-images.githubusercontent.com/15953087/236762812-33692ed3-f1d3-4b8a-9e38-ff836ae8245b.png)

**Comment by alexander-akait** [comment]
Do you really have `RemoteComponent` in `app1`?

**Comment by tony-show** [comment]
> Do you really have `RemoteComponent` in `app1`?

**Yes**
![image](https://user-images.githubusercontent.com/15953087/237035667-1a4300df-41f3-401a-b5b9-f10b9276056f.png)

**Comment by tony-show** [comment]
> @tony-show Sorry for delay in answers, can I use your repo to check it and what commands do you run (sorry for these stupid questions, there are a lot of issues, so hard to track and remember all of them)?

Hello Alexander!
Yes, of course you can use this repository! This is a test repository.

**Comment by alexander-akait** [comment]
@tony-show I will look soon, sorry for delay, a lot of issues

**Comment by anatoly-ivashov** [comment]
@alexander-akait Hi! Is there any news about fixing the problem?

**Comment by alexander-akait** [comment]
@anatoly-ivashov No, still in my TODO, try looking at it this week

**Comment by anatoly-ivashov** [comment]
@alexander-akait Thank you for your answer. I hope you can find some time to find a solution to this problem. Anyway, thanks Alexander, for your contribution to many useful things for it community!

**Comment by WangZhenHao** [comment]
I encountered this problem too.But when I migrate to Vite，The Hot reload can work!

**Comment by kujtimprenku** [comment]
Did anyone manage to get this sorted? - It seems to happen when the host-app `exposes` anything.

**Comment by WangZhenHao** [comment]
> Did anyone manage to get this sorted? - It seems to happen when the host-app `exposes` anything.

![image](https://github.com/webpack/webpack-dev-server/assets/26181934/6bdccb1e-e26a-4341-8d96-3a208186d42e)

I commented that switching to using Vite might solve your current preblom, But the Administrator deleted my comments. why?

**Comment by kujtimprenku** [comment]
> > Did anyone manage to get this sorted? - It seems to happen when the host-app `exposes` anything.
> 
> ![image](https://private-user-images.githubusercontent.com/26181934/339178787-6bdccb1e-e26a-4341-8d96-3a208186d42e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyODA3MzgsIm5iZiI6MTcxODI4MDQzOCwicGF0aCI6Ii8yNjE4MTkzNC8zMzkxNzg3ODctNmJkY2NiMWUtZTI2YS00MzQxLThkOTYtM2EyMDgxODZkNDJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEzVDEyMDcxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJmYzM5YjkwZDAxOWRjYmExNDkwNjRkMjljYmVlZjg1MzVkN2EyMjliMGYzOTM5M2NmMDU3MzA0NGFlYzVmMjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._ETJDbK3qLkbY6CZK7HGMDm06GW6edLSNGJlCy8EiGg)
> 
> I commented that switching to using Vite might solve your current preblom, But the Administrator deleted my comments. why?

Thanks for the reply, unfortunately, switching to another bundler is not an option for this project I'm working on.
Since this happens only on host-app (shell) and we rarely change it, I can live with that.

**Comment by igorpodosonov** [comment]
A temporary workaround that has worked for me involves disabling HMR and enabling live reloading. You can adjust your devServer configuration as follows:

```
devServer: {
  hot: false, // Disable hot module replacement
  liveReload: true, // Enable full page reloads when changes are detected
}
```

**Comment by lobaak** [comment]
Is there still no fix for this issue?

**Comment by jwq403926021** [comment]
I meet the same issue. 
[https://github.com/module-federation/core/issues/3289](https://github.com/module-federation/core/issues/3289)

**Comment by jwq403926021** [comment]
@alesculek 
Try to set NODE_ENV=development as your env variable. then start the project.
I met the same error recently. https://github.com/module-federation/core/issues/3289
